dimod/include/dimod/constrained.h

```diff
@@ -0,0 +1,320 @@
+ private:
+    const parent_type* parent_;
+
+    std::vector<index_type> variables_;
+    Sense sense_;
+    bias_type rhs_;
+
+    std::unordered_map<index_type, index_type> indices_;
+
+ public:
+    Constraint(const parent_type* parent, Sense sense)
+            : base_type(), parent_(parent), variables_(), sense_(sense), rhs_(0) {}
+
+    Constraint(const parent_type* parent, std::vector<index_type>&& variables,
+               std::vector<bias_type>&& biases, Sense sense, bias_type rhs)
+            : base_type(std::move(biases)),
+              parent_(parent),
+              variables_(std::move(variables)),
+              sense_(sense),
+              rhs_(rhs) {
+        index_type i = 0;
+        for (const index_type& v : this->variables_) {
+            this->indices_[v] = i++;
+        }
+    }
+
+    Constraint(const parent_type* parent, const Constraint& constraint)
+            : base_type(constraint),
+              parent_(parent),
+              variables_(constraint.variables_),
+              sense_(constraint.sense_),
+              rhs_(constraint.rhs_),
+              indices_(constraint.indices_) {}
+
+    Constraint(Constraint&& other) noexcept { *this = std::move(other); }
+
+    Constraint& operator=(Constraint&& other) noexcept {
+        if (this != &other) {
+            base_type::operator=(std::move(other));
+            this->parent_ = other.parent_;
+            this->variables_ = std::move(other.variables_);
+            this->sense_ = other.sense_;
+            this->rhs_ = other.rhs_;
+            this->indices_ = std::move(other.indices_);
+        }
+        return *this;
+    }
+
+    void add_linear(index_type v, bias_type bias) {
+        base_type::add_linear(this->add_variable(v), bias);
+    }
+
+    void add_quadratic(index_type u, index_type v, bias_type bias) {
+        base_type::add_quadratic(this->add_variable(u), this->add_variable(v), bias);
+    }
+
+    [[noreturn]] void add_quadratic_back(index_type u, index_type v, bias_type bias) {
+        throw std::logic_error("not implemented");
+    }
+
+    template <class T>
+    void fix_variable(index_type v, T assignment) {
+        throw std::logic_error("todo 110");
+    }
+
+    bias_type linear(index_type v) const {
+        assert(v >= 0 && static_cast<size_type>(v) < this->parent_->num_variables());
+        auto it = this->indices_.find(v);
+        if (it == this->indices_.end()) {
+            return 0;
+        }
+        return base_type::linear(it->second);
+    }
+
+    bias_type lower_bound(index_type v) const { return this->parent_->lower_bound(v); }
+
+    bias_type rhs() const { return this->rhs_; }
+
+    Sense sense() const { return this->sense_; }
+
+    void set_linear(index_type v, bias_type bias) {
+        base_type::set_linear(this->add_variable(v), bias);
+    }
+
+    void set_quadratic(index_type u, index_type v, bias_type bias) {
+        base_type::set_quadratic(this->add_variable(u), this->add_variable(v), bias);
```

Comment:
I have something in c++. You could improve the performance here by templating the function add_variable, so that in cases where you know the variable should be present or not present, you can pass in the template variable to call the corresponding generated function.

Reply:
Cool, will look into it.